handler/src/main/java/io/netty/handler/ssl/AbstractSniHandler.java

```diff
@@ -0,0 +1,323 @@
+                                }
+                            }
+                            // Fall-through
+                        default:
+                            //not tls, ssl or application data, do not try sni
+                            break loop;
+                    }
+                }
+            } catch (Throwable e) {
+                // unexpected encoding, ignore sni and use default
+                if (logger.isDebugEnabled()) {
+                    logger.debug("Unexpected client hello packet: " + ByteBufUtil.hexDump(in), e);
+                }
+            }
+            // Just select the default SslContext
+            select(ctx, null);
+        }
+    }
+
+    private void select(final ChannelHandlerContext ctx, final String hostname) throws Exception {
+        Future<T> future = lookup(ctx, hostname);
+        if (future.isDone()) {
+            onLookupComplete(ctx, hostname, future);
+        } else {
+            suppressRead = true;
+            future.addListener(new FutureListener<T>() {
+                @Override
+                public void operationComplete(Future<T> future) throws Exception {
+                    try {
+                        suppressRead = false;
+                        try {
+                            onLookupComplete(ctx, hostname, future);
+                        } catch (DecoderException err) {
+                            ctx.fireExceptionCaught(err);
+                        } catch (Throwable cause) {
+                            ctx.fireExceptionCaught(new DecoderException(cause));
+                        }
+                    } finally {
+                        if (readPending) {
+                            readPending = false;
+                            ctx.read();
+                        }
+                    }
+                }
+            });
+        }
+    }
+
+    /**
+     * Kicks off a lookup for the given SNI value and returns a {@link Future} which in turn will
+     * notify the {@link #onLookupValue(ChannelHandlerContext, String, Future)} on completion.
+     *
+     * @see #onLookupValue(ChannelHandlerContext, String, Future)
+     */
+    protected abstract Future<T> lookup(ChannelHandlerContext ctx, String hostname) throws Exception;
+
+    /**
+     * Called upon completion of the {@link #lookup(ChannelHandlerContext, String)} {@link Future}.
+     *
+     * @see #lookup(ChannelHandlerContext, String)
+     */
+    protected void onLookupComplete(ChannelHandlerContext ctx, String hostname, Future<T> future) throws Exception {
+        if (!future.isSuccess()) {
+            throw new DecoderException("failed to get the SslContext for " + hostname, future.cause());
+        }
+
+        onLookupValue(ctx, hostname, future.getNow());
+    }
+
+    /**
+     * @see #onLookupComplete(ChannelHandlerContext, String, Future)
+     */
+    protected abstract void onLookupValue(ChannelHandlerContext ctx, String hostname, T value) throws Exception;
```

Comment:
I've removed `onLookupValue()` and made `onLookupComplete` an abstract method. All biz logic lives in `SniHandler` and that's the class people should use.

For people like me, I'm only interested in the meat that is provided by the `decode(...)` method. I also prefer to work with the `Future<T>`[1] instead of just `T`. Hence `onLookupComplete(..., Future<T>)`.

[1] E.g. I do have a version of Netty Future/Promise impl. that implements JDKs CompletionStage and when I extend AbstractSniHandler I maybe want to do something like this...

```java
@Overrides
protected void onLookupComplete(...., Future<SslContext> future) throws Exception {
  CompletableFuture<SslContext> f = (CompletableFuture<SslContext>)future;
  // ... do my own thing ...
}
```